pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/unsafe/MmapMemory.java

```diff
@@ -0,0 +1,348 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.spi.memory.unsafe;
+
+import com.google.common.collect.Lists;
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.io.UncheckedIOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.channels.FileChannel;
+import java.util.List;
+import java.util.function.BiConsumer;
+import net.openhft.chronicle.core.Jvm;
+import net.openhft.chronicle.core.OS;
+import net.openhft.posix.MSyncFlag;
+import net.openhft.posix.PosixAPI;
+import org.apache.pinot.segment.spi.utils.JavaVersion;
+
+
+public class MmapMemory implements Memory {
+
+  private static final MapFun MAP_FUN;
+
+  /**
+   * The address actually mapped. It has to be page aligned.
+   *
+   * {@code _address = _offset - offset % pageSize}
+   */
+  private final long _address;
+  /**
+   * The offset requested to map.
+   *
+   * {@code _address = _offset - offset % pageSize}
+   */
+  private final long _offset;
+  /**
+   * How many bytes have been requested to be mapped.
+   * The actual mapped size may be larger (up to the next page), but the actual mapped size
+   * is stored by {@link #_section}.
+   */
+  private final long _size;
+  private final MapSection _section;
+
+  static {
+    try {
+      Jvm.init();
+      MAP_FUN = MapFun.find();
+    } catch (ClassNotFoundException | NoSuchMethodException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public MmapMemory(File file, boolean readOnly, long offset, long size) {
+    _size = size;
+    _offset = offset;
+
+    try {
+      _section = MAP_FUN.map(file, readOnly, offset, size);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    _address = _section.getAddress();
+  }
+
+  @Override
+  public long getAddress() {
+    return _address;
+  }
+
+  @Override
+  public long getSize() {
+    return _size;
+  }
+
+  @Override
+  public void flush() {
+    MSyncFlag mode = MSyncFlag.MS_SYNC;
+    PosixAPI.posix().msync(_offset, _size, mode);
```

Comment:
Seems quite difficult to find which unsafe method we should call and they change on each JVM version. Therefore this idea of calling the actual posix method seems better and it works in both mac and linux.

Reply:
I saw here and JVM versions are the two places where openhft dependencies got introduced.

I feel if we only want to support limit java LTS versions, then it's ok to have the implementations per jdk version.

Reply:
This is not a blocker for java 17 upgrade effort, we can revisit this later.

Reply:
In order to do the flush, it seems better to use PosixAPI library. I've tried to do the same trick I made with map and unmap, but the private Java methods seem to be quite different between JVMs and require things like having the file descriptor, which is not easily available in Java and it seems we should do again the same reflection trick to get the `fd`.

But I have to be honest, I didn't dedicate too much time here. Once I've seen that it looks complicated, I look for alternatives and once I found the Posix library, I opted in to use that alternative. 

BTW, There is a third option/. Pinot already have a dependency to `com.github.jnr:jnr-posix`. If that library let us call this posix function, we wouldn't need to add the openhft dependency.